llvm/lib/Target/SPIRV/SPIRVGlobalTypesAndRegNumPass.cpp

```diff
@@ -731,6 +750,121 @@ static void addGlobalRequirements(const SPIRVRequirementHandler &reqs,
+                                   const SPIRVSubtarget &ST) {
+  auto *DT = ST.getSPIRVDuplicatesTracker();
+  auto *TR = ST.getSPIRVTypeRegistry();
+  auto &MF = MIRBuilder.getMF();
+
+  // Walk over each MF's DT and select all unreferenced global variables.
+  SmallVector<GlobalVariable *, 8> GlobalVarList;
+  const SPIRVDuplicatesTracker<GlobalValue> *GVDT = DT->get<GlobalValue>();
+  auto map = GVDT->getAllUses();
+  for (Module::global_iterator I = M.global_begin(), E = M.global_end();
+       I != E;) {
+    GlobalVariable *GV = &*I++;
+    auto addrSpace = GV->getAddressSpace();
+    auto storage = TR->addressSpaceToStorageClass(addrSpace);
+    if (storage != StorageClass::Function) {
+      auto t = map.find(GV);
+      if (t == map.end())
+        // GV is not found in maps of all MFs so it's unreferenced.
+        GlobalVarList.push_back(GV);
+    }
+  }
+
+  // Walk over all unreferenced global variables and create required types,
+  // constants, OpNames, OpVariables and OpDecorates.
+  TR->setCurrentFunc(MF);
+  setMetaBlock(MIRBuilder, MB_TmpGlobalData);
+  for (GlobalVariable *GV : GlobalVarList) {
+    auto addrSpace = GV->getAddressSpace();
+    auto storage = TR->addressSpaceToStorageClass(addrSpace);
+
+    Register Reg;
+    if (storage != StorageClass::Function &&
+        DT->find(GV, &MF, Reg) == false) {
+      auto LLVMTy = GV->getType();
+      Reg = MIRBuilder.getMRI()->createGenericVirtualRegister(LLT::scalar(32));
+      SPIRVType *spvTy = TR->assignTypeToVReg(LLVMTy, Reg, MIRBuilder);
+      DT->add(GV, &MF, Reg);
+
+      Register initVReg = 0;
+      if (GV->hasInitializer()) {
+        Constant *InitVal = GV->getInitializer();
+        SPIRVType *initTy = TR->getOrCreateSPIRVType(InitVal->getType(),
+                                                     MIRBuilder);
+        if (const ConstantInt *CI = dyn_cast<ConstantInt>(InitVal))
+          initVReg = TR->buildConstantInt(CI->getValue().getZExtValue(),
+                                          MIRBuilder,  initTy);
+        else if (const ConstantFP *CF = dyn_cast<ConstantFP>(InitVal))
+          initVReg = TR->buildConstantFP(CF->getValue(),
+                                         MIRBuilder, initTy);
+      }
+
+      if (GV->getLinkage() == GlobalValue::LinkageTypes::ExternalLinkage) {
+        auto globalIdent = GV->getGlobalIdentifier();
+        auto MIB = MIRBuilder.buildInstr(SPIRV::OpDecorate)
+                   .addUse(Reg)
+                   .addImm(Decoration::LinkageAttributes);
+        addStringImm(globalIdent, MIB);
+        MIB.addImm(LinkageType::Export);
+      }
+
+      auto MIB = MIRBuilder.buildInstr(SPIRV::OpVariable)
+               .addDef(Reg)
+               .addUse(TR->getSPIRVTypeID(spvTy))
+               .addImm(storage);
+      if (initVReg)
+        MIB.addReg(initVReg);
+      TR->constrainRegOperands(MIB);
+    }
+  }
+
+  // Walk over all created instructions and add requirement. Also convert all
```

Comment:
I suppose those constants should've been processed by InstructionSelector, why that didn't happen? Were they created after it?

Reply:
These constants were created just now in the loop at 791 line (walk unreferenced global vars). We can handle them only in a module pass. Currently  I don't think that we need to introduce one more module level pass before InstructionSelector to process these global vars, so I put its creation and processing here. It's not very good design but may be acceptable for now if we understand and control what happens here.